lm_eval/tasks/ja/jaqket_v2.py

```diff
@@ -63,24 +79,54 @@ def validation_docs(self):
             dataset = [item for item in dataset if item["id"] not in self.REMOVE_IDS]
         return dataset
     
+    def doc_to_qa_prompt(self, doc):
+        return (
+            "[質問]:"
+            + doc["question"]
+            + self.SEP
+            + "[答え]:"
+        )
+
     def doc_to_text(self, doc):
         topk_titles = doc["ctxs"]["title"][:self.TOP_K_LIMIT]
         topk_contexts = doc["ctxs"]["text"][:self.TOP_K_LIMIT]
-        context = f"{self.SEP}".join([
-            "[題名]:"
-            + title
-            + f"{self.SEP}"
-            + "[問題]:"
-            + context
+        answer_candidate = self.SEP.join([
+            (
+                "[題名]:"
+                + title
+                + self.SEP
+                + "[問題]:"
+                + context
+            )
             for title, context in zip(topk_titles, topk_contexts)
         ])
+        qa_prompt = self.doc_to_qa_prompt(doc)
         return (
-            context
-            + f"{self.SEP}"
-            + "[質問]:"
-            + doc["question"]
-            + f"{self.SEP}"
-            + "[答え]:"
+            answer_candidate
+            + self.SEP
+            + qa_prompt   
+        )
+
+    def doc_to_answering_text(self, doc, fallback_doc):
+        answering_contexts = [
+            {k: v[i] for k, v in doc["ctxs"].items()}
+            for i, a in enumerate(doc["ctxs"]["has_answer"]) if a == True
+        ]
+        if len(answering_contexts) < 1:
```

Comment:
How many times is the fallback doc called? There's only one fallback and that might lead to lack of randomness. For example, a model is very good at that fallback and it's called multiple times. Then the score would be biased.

Reply:
I think this is OK under the assumption this isn't called very much. I'm not sure what it would mean for the model to be good at the fallback - in that case it would just be a like a special prompt, which should be reported, but isn't leakage or anything.

I also had two separate concerns about this code.

First, when is `has_answer` not `True`? It's not clear to me why there would be any question like that in the dataset.

Second, I find this code hard to follow. You're iterating over `doc["ctxs"]` as a dictionary, but in the fallback doc it's a list. What is the actual structure of this object? Can you give an example in a comment?

Reply:
regarding to the first point, `has_answer=False` existed due to slicing contexts by `TOP_K_LIMIT` .
but now the task does slicing by top k and filtering by `has_answer` on `load_dataset()`.

and, the second point, I dropped iterating over doc["ctxs"] and fallback doc totally.